lib/blocks.php

```diff
@@ -296,3 +296,189 @@ function gutenberg_enqueue_editor_block_styles_assets() {
+ * which will be applied to the block markup in the front-end.
+ *
+ * @param  array $attributes block attributes.
+ * @return array Colors CSS classes and inline styles.
+ */
+function gutenberg_experimental_build_css_colors( $attributes ) {
+	$text_colors       = array(
+		'css_classes'   => array(),
+		'inline_styles' => '',
+	);
+	$background_colors = array(
+		'css_classes'   => array(),
+		'inline_styles' => '',
+	);
+
+	// Text color.
+	$has_named_text_color  = array_key_exists( 'textColor', $attributes );
+	$has_custom_text_color = array_key_exists( 'style', $attributes )
+	&& array_key_exists( 'color', $attributes['style'] )
+	&& array_key_exists( 'text', $attributes['style']['color'] );
+
+	// If has text color.
+	if ( $has_custom_text_color || $has_named_text_color ) {
+		// Add has-text-color class.
+		$text_colors['css_classes'][] = 'has-text-color';
+	}
+
+	if ( $has_named_text_color ) {
+		// Add the color class.
+		$text_colors['css_classes'][] = sprintf( 'has-%s-color', $attributes['textColor'] );
+	} elseif ( $has_custom_text_color ) {
+		// Add the custom color inline style.
+		$text_colors['inline_styles'] .= sprintf( 'color: %s;', $attributes['style']['color']['text'] );
+	}
+
+	// Link colors.
+	$has_link_color = array_key_exists( 'style', $attributes )
+	&& array_key_exists( 'color', $attributes['style'] )
+	&& array_key_exists( 'link', $attributes['style']['color'] );
+
+	if ( $has_link_color ) {
+		$text_colors['css_classes'][] = 'has-link-color';
+		// If link is a named color.
+		if ( strpos( $attributes['style']['color']['link'], 'var:preset|color|' ) !== false ) {
+			// Get the name from the string and add proper styles.
+			$index_to_splice               = strrpos( $attributes['style']['color']['link'], '|' ) + 1;
+			$link_color_name               = substr( $attributes['style']['color']['link'], $index_to_splice );
+			$text_colors['inline_styles'] .= sprintf( '--wp--style--color--link:var(--wp--preset--color--%s);', $link_color_name );
+		} else {
+			$text_colors['inline_styles'] .= sprintf( '--wp--style--color--link: %s;', $attributes['style']['color']['link'] );
+		}
+	}
+	// Background color.
+	$has_named_background_color  = array_key_exists( 'backgroundColor', $attributes );
+	$has_custom_background_color = array_key_exists( 'style', $attributes )
+	&& array_key_exists( 'color', $attributes['style'] )
+	&& array_key_exists( 'background', $attributes['style']['color'] );
+
+	// Gradient color.
+	$has_named_gradient  = array_key_exists( 'gradient', $attributes );
+	$has_custom_gradient = array_key_exists( 'style', $attributes )
+	&& array_key_exists( 'color', $attributes['style'] )
+	&& array_key_exists( 'gradient', $attributes['style']['color'] );
+
+	// If has background color.
+	if ( $has_custom_background_color || $has_named_background_color || $has_named_gradient || $has_custom_gradient ) {
+		// Add has-background-color class.
+		$background_colors['css_classes'][] = 'has-background-color';
```

Comment:
We probably want to change this to just 'has-background'.  
related - https://github.com/WordPress/gutenberg/pull/23044#discussion_r438608423